velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -280,4 +280,43 @@ struct CombinationsFunction {
   }
 };
 
+template <typename T>
+struct udf_array_sum {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+  template <typename OT, typename IT>
+  FOLLY_ALWAYS_INLINE bool call(OT& out, const IT& array) {
+    if (array.mayHaveNulls()) {
+      bool allNulls = true;
+      OT sum = 0;
+      for (const auto& item : array) {
+        if (item.has_value()) {
+          allNulls = false;
+          sum += *item;
+        }
+      }
+      out = sum;
+      return true;
+    }
+
+    // Not nulls path
+    OT sum = 0;
+    for (const auto& item : array) {
+      sum += *item;
+    }
+    out = sum;
+    return true;
+  }
+
+  template <typename OT, typename IT>
+  FOLLY_ALWAYS_INLINE bool callNullFree(OT& out, const IT& array) {
```

Comment:
What's the motivation for defining both "call" and "callNullFree"?

Reply:
The callNullFree method removes the null check on each element in the array and has better performance when there aren't nulls in the input.

Reply:
Sure, but then you don't need 'call' method. See @laithsakka comments as well.

Reply:
Shouldn't the call method be used when there are nulls in the input? It's required because without it the default null behavior would be used and any input with nulls would have the result set to null and not to zero as the specification requires.

Reply:
I didn't fully understand how callNullFree behaves. I read the docs and now understand that you need 'call' function, but callNullFree is an optional optimization for no-nulls case.

https://facebookincubator.github.io/velox/develop/scalar-functions.html#null-free-fast-path

Reply:
@mbasmanova  @jwyles-ahana @kevinwilfong 
callNullFree works as the following:
1. if CallNullFree exists we will check if there is nulls recursively if any null is found there are two possibilities
 - if a call or callNullable exists then we call that function. 
 - otherwise, we return null. (

Reply:
@laithsakka Thank you for clarifying.